### DISCUSSION TITLE
Workplan for v0.3.0 -- Tell us what you think / want!

### DISCUSSION THREAD

**Discussion by vmoens** [pytorch/rl#1612] [Ideas]
Now that v0.2.0 is out, let's talk v0.3.0!

## Refactoring

- [ ] 
- [ ] NOT PLANNED ~Removing trailing dim in done, reward, etc
The extra singleton dim is annoying, unclear if we should use it for _reset etc.
The main motivation to have it was to avoid silent bugs like `reward + (1-terminated) * next_val * gamma` but these things are internal and we already check the shapes so we can expect that where things will break, they will be easy to spot.
This **will** be bc-breaklng, I don't see a clear path to smoothly change it (maybe a flag like old_api or smth but that isn't the greatest as we will need to do a lot of assumptions in losses and similar).~
- [x] `env.step_and_maybe_reset` for faster data collection
- [x] Refactoring stateless models in losses for better compat, possibly with torch.func.functional_call

## New algos and fixes
- [x] IMPALA, and possibly DDPPO
- [ ] Improving RLHF primitives => v0.4
- [ ] Fix dreamer => v0.4 @BY571 
- [x] New datasets (see #509)

## Improvements
- [x] Avoid loading gym with `set_gym_backend` and similar functions
- [x] Faster data collection on CUDA (#1325):
  - [x] different device for env and sub-envs in ParallelEnv #1626 
  - [x] `policy_device`, `env_device` in collector (see below)
- [ ] GRU / LSTM compatible with tricky data structures => v0.4
- [x] Proper way of saving memmap replay buffers (#1588)

**Comment by vmoens** [comment]
# Refactoring of losses

## Deprecate usage of tensordict functional calls

As noted in #1613, functional calls like `module(*input, params)` can cause some problems if the module cannot be deepcopied. There are many other issues with this, for instance recognising the `params` input with any given signature can be quite hard. 
Another option could be to use `torch.func.functional_call` but, as of today, this does not support `TensorDict` inputs well (i will if https://github.com/pytorch/pytorch/pull/112441 goes through but I'm keeping my expectations low). 
The option I would go for is to use this feature from `TensorDict` that we will land soon:
```python
with params.to_module(module):
    output = module(input)
```
which (1) works well with tensordict and (2) allows us to call other methods than forward (eg, `get_dist`) and (3) allows for more complex calls like `vmap` or `grad` within the decorator (@matteobettini).

**What you should expect**: In theory this won't be bc-breaking, all changes will occur within the losses. We'll keep `make_functional` and related for backward compatibility but losses won't use these anymore. From a performance point of view, I wouldn't expect any substential slowdown. 

## Modularizing losses

Some losses (eg, TD3) already benefit from methods like `actor_loss` and `value_loss` that can be called indepdently. The forward method just calls them one after another and groups the results. We'd like to generalize this to all losses. In some cases (eg, CQL) this can even allow users to quickly make ablation tests (cc @BY571).

## `reduction` arg in losses

It has been brought to our attention that calling `mean` on losses by default may be unwanted in some cases. We could add a `reduction` kwarg in constructors to allow for different reduction strategies. cc @albertbou92

**Comment by vmoens** [comment]
On the data collector side, we could make things run faster by allowing a more flexible device handling.

Currently, the device of data collectors is poorly handled IMO. It's not super clear what your env and policy run on if you pass or don't pass a device.
Ultimately, I think the best API would be to be the following:
- if you do not specify any device during construction, the collector uses the env and policy device and returns data on the env device (because we need to pick one between policy and env, but policy does not generally have a `.device` attribute). All collectors will perform the casting from device to device internally.
- If you want, you can cast policy and env to a specific device by using `policy_device` and `env_device` kwargs. 
- We keep the `storing_device` which indicates where the data needs to be stored (if you have very long rollout you may want to dump data on RAM). So we'll have 3 devices that you can pass to the constructor.
- Finally, the current `device` will be kept for backward compatibility purposes and will override `policy_device` and `env_device` (an error will be thrown if you pass both `device` and one of `env_device` and `policy_device`).

## Will this be bc-breaking

I believe not

## What to expect from these changes

The main thing I want to resolve is that executing envs that would naturally be on CPU with a parallel env on cuda is slow because there is a lot of io.
With this change, you will be able to execute your envs on CPU, and automatically cast the data on CUDA to run your policy. That will be much faster than having a parallel env on cuda! (cc @skandermoalla)

cc @matteobettini this is something we talked about I believe

**Comment by matteobettini** [comment]
Thanks a lot for this discussion. Alll the features listed are going to be incredibly helpful for our lab as users of TorchRL, and I agree with all the roadmaps set.

A question I had is about the defaults for the new device arguments passed to collectors. I imagine the policy and env devices will default to the current device of these compnents. But what will be the default for `storing_device`?

**Reply by vmoens** [reply]
I would stand by 
> returns data on the env device (because we need to pick one between policy and env, but policy does not generally have a .device attribute)

**Comment by skandermoalla** [comment]
What is the plan for improving RLHF primitives? Can one do distributed RLHF with TorchRL across multiple nodes already?

**Reply by vmoens** [reply]
What I think would be the most beneficial is to be able to train models at scale.
Right now our functional API isn't the best.
My ultimate goal would be to have a couple of distributed models that you can call functionally with 
```python
with params.to_module(module): y = module(x)
```
Where module is a DDP model for instance, or params contains shared tensors (DTensors).

That would permit training big models with TorchRL.

Other than that we should improve our transforms and the way we play with them.

Last item on my list for this release is better usage of TensorDict throughout the code, eg using TensorDict.map for transformation. A better interaction with test data would be highly valuable too.

For v0.4 I think a better integration within the trainer class would be helpful.

**Reply by wbinventor** [reply]
Is it currently possible to train large models with DDP with TorchRL? For example, can the `LossModule` be instantiated with DDP models? I've recently run into an issue when trying to do this, and came across this comment when trying to understand how to use TorchRL with DDP.

**Reply by vmoens** [reply]
Currently our reliance on functional calls makes this hard. 
In some instances (eg, naive PPO usage), functional calls are not needed though, hence we could spare users from relying on it and this would make it possible to use DDP or other distributed backends out of the box.
Functional calls are needed whenever one does meta-rl (not frequent) or vmaps over sets of parameters (eg, multiple Q-value networks => quite frequent). We could loop over the parameters if the module is DDP but the code will soon become unreadable.
Anyhow: I'm just throwing things here to give you an idea of why that is going to be challenging but it is 100% sure that this is something we need so we will do whatever is needed to make it happen!
I am planning on exploring these avenues this very month, I will update you on the progress!

**Reply by wbinventor** [reply]
> I am planning on exploring these avenues this very month, I will update you on the progress!

That sounds wonderful, I'm very interested to follow any updates on this front!

In the meantime, would be great to know if there is any workaround to run e.g., naive PPO with DDP models.

**Reply by vmoens** [reply]
Tell me what objective you want and I can try to make it happen

**Reply by wbinventor** [reply]
As an example, consider actor and critic models that are each `nn.Module`s wrapped with [`DistributedDataParallel`](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel). Then `TensorDictModule`s are created from each of these and used to instantiate the `PPOLoss`. At least for my code, this currently fails in the `deepcopy` in `convert_to_functional`. For simplicity, assume that each distributed process may separately initialize an environment, replay buffer, etc. (without using any of TorchRL's fancier distributed/sync/async primitives).

Let me know if this rough outline makes sense or not. I'll see if I can reproduce this error in a simple example script derived from one of the TorchRL tutorials but where the models are `DistributedDataParallel`.

**Reply by wbinventor** [reply]
@vmoens I've found that `LossModule().convert_to_functional` can be made to work with DDP models _if_ the `LossModule` is created before wrapping the model with [DistributedDataParallel](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel). Without this, I run into an error on the `copy.deepcopy(module)` called by `convert_to_functional`.

Alternatively, if my model's `__deepcopy__` operator is (temporarily) made to forward to a shallow `copy.copy`, then the `LossModule` not only works for DDP models, but also even [`FullyShardedDataParallel`](https://pytorch.org/docs/stable/fsdp.html#module-torch.distributed.fsdp)-wrapped models.

Just curious if you would recommend against either of these approaches to initialize a `LossModule` with DDP/FSDP-wrapped models, or if these workarounds might have any unintended consequences?

**Reply by vmoens** [reply]
I would need to test them first, do you mind if i dedicate more time to this next week?

**Reply by wbinventor** [reply]
That sounds great! Definitely interested to hear if this is an appropriate workaround, for now.

**Reply by vmoens** [reply]
@wbinventor https://github.com/pytorch/rl/pull/1804
This is a proposal for PPO, A2C and Reinforce.
I have an idea of how to make the others also functionalless but that will take a bit more time!

**Reply by wbinventor** [reply]
@vmoens awesome, I will pull and test this out!

**Reply by wbinventor** [reply]
This patch looks good and works for me 😄

**Comment by wbinventor** [comment]
When will v0.3.0 be released? And will it come with the same requirement on PyTorch >=2.2.0 that `torchRL-nightly` has?

**Reply by vmoens** [reply]
Release date at the end of the month
RE pytorch requirement: if you pip install yes it will ask for the latest torch. But if you make a local install through a git clone, anything above torch 1.13 will work
